src/lucky/mime_type.cr

```diff
@@ -0,0 +1,94 @@
+# :nodoc:
+class Lucky::MimeType
+  alias Format = Symbol
+  alias AcceptHeaderSubstring = String
+  class_getter accept_header_formats = {} of AcceptHeaderSubstring => Format
+
+  register "application/json", :json
+  register "text/json", :json
+  register "application/jsonrequest", :json
+  register "application/xml", :xml
+  register "application/rss+xml", :rss
+  register "application/atom+xml", :atom
+  register "application/x-yaml", :yaml
+  register "text/yaml", :yaml
+  register "text/csv", :csv
+  register "text/css", :css
+  register "text/calendar", :ics
+  register "text/plain", :plain_text
+  register "multipart/form-data", :multipart_form
+  register "application/x-www-form-urlencoded", :url_encoded_form
+
```

Comment:
Does `:html` need to be in this list? Also, maybe `"text/javascript", :js`?

Reply:
`:html` is not really reliable since browser use a super weird default header (I think). It seems like what other frameworks do is they fallback to HTML if no accept header is given and :html is the default. But I will research this more to make sure we are doing this right.

Reply:
I like the js idea. I'll add that

Reply:
Phoenix docs: https://hexdocs.pm/phoenix/1.4.9/Phoenix.Controller.html#accepts/2

> It is important to notice that browsers have historically sent bad accept headers. For this reason, this function will default to "html" format whenever:
>
> * the accepted list of arguments contains the "html" format
>
> * the accept header specified more than one media type preceded or followed by the wildcard media type "*/*"

I think this will work if you have `accepted_formats [:html, :json], default: :html` *but* if you do something like `accepted_formats [:html, :json], default: :json` then `:html` will never match...I wonder how to handle this case. I wonder if we should force defaulting to `:html` if you accept it? Or maybe we need to do something in Lucky where if there is a `*/*` and there is no accept header and `:html` is accepted we use that?

This PR has been hard lol

Reply:
Yeah, I've seen the same with HTML. I think that being the default fallback is fine since this is a web framework, and html is at least the 60% use case 😂

Reply:
Browsers will also send `text/html` inside of the accept headers. If you have default `:json`, but the accept header has `*/*` and `text/html`, then it's html.

Reply:
But only if you also accept `:html`. If you only accept `:json` then you'd want to default to `:json` and not assume `:html`. That's part of why this is such a PITA. So many edge cases. Lol

I think what we can do is if you have `:html` in accepted formats, you *must* have `:html` as the default. I think this makes sense anyway since if you set `:json` or something, how would a use using a browser request html instead?

It's a little weird, but I think it makes sense to just force that and raise a nice error telling people why that is if they try to default to something different. Thoughts?

Reply:
Just to be clear, we're talking about when a request comes in that doesn't have any accept header? Like `curl http://localhost:3001/posts.json`. We have `[:html, :json]`, but if `:html` is default, then that's what would be returned here? That actually might be ok. If we do that, then [this pr](https://github.com/luckyframework/lucky/pull/778) can be re-worked a little to see that `.json` file extension, set the proper accept header, and then the issue wouldn't matter at that point.

Reply:
Let's say you have an `ApiAction` with `accepted_formats [:json]`

If we made it so that the browser header sets html or `text/html` sets html, then when you visit any ApiAction from the browser it would throw an error. You *could* use the trailing format, but I think as a user I would expect/want it to just give me JSON since that's the only thing it accepts.

So we can't just set `:html` format if it includes `text/html` or `*/*`. We have to fallback to the default if the header is a browser default `*/*` (in this case, :json).

However that breaks down if someone does `accepted_formats [:html, :json], default: :json` because then the browser headers will be `*/*` and would set the format to `:json`. Meaning the format would never be `:html` :( So what I think we should do is force the default to `:html` if you accept `:html`. The downside is if you curl that endpoint without a trailing format or a `application/json` you'd get back HTML. I think that is fine. You should specify if you want something other than the default

Does that make sense? Did I understand your question/thoughts correctly?

Reply:
I'm glad we're talking about this because it is a huge PITA and hard to get right :P

Reply:
Actually I'm prob overthinking this. We should just do what Phoenix does. If we have a browser default header, and the action accepts :html, then we use :html. Otherwise use the default

Reply:
Ah, yeah. I've never had to hit an api from the browser, but that brings up a good point. It could be more than just json. You might hit a CSV endpoint from the browser and such. I don't really see another way around aside from your suggestion. Let's just roll with that for now, and if we come up with a better idea after playing with it for a bit, then we can always make updates with new releases.